### PR TITLE
Feature/nuc 2325 create vuexp cli plugin

### PR DESCRIPTION
**Pull request by osmanertem**

### CODE REVIEW

File: generator/index.js [hunk 1]
```diff
@@ -0,0 +1,771 @@
...
+      'clean:android': 'rimraf platforms/android',
+      'clean:ios': 'rimraf platforms/ios'
+    },
+    dependencies: {
+      'nativescript-vue': '^2.2.2',
+      vuexp: '^0.3.0',
+      'vuexp-router': 'git://github.com/vuexp/vuexp-router.git#development',
+      'tns-core-modules': '^5.4.1',
+      'nativescript-background-http': '^3.4.0',
+      'nativescript-imagepicker': 'git://github.com/mylmz10/nativescript-mediapicker.git',
```
Comment (brendtumi) [inline generator/index.js]: @mylmz10 you need to publish this library. Using git path is not secure.

File: generator/index.js [hunk 2]
```diff
@@ -0,0 +1,771 @@
...
+      'preview:ios': 'npm run setup-webpack-config && tns preview --bundle --env.development --env.ios',
+      'setup-webpack-config': 'node ./node_modules/vue-cli-plugin-vuexp/lib/scripts/webpack-maintenance pre',
+      'clean:platforms': 'rimraf platforms',
+      'clean:android': 'rimraf platforms/android',
+      'clean:ios': 'rimraf platforms/ios'
+    },
+    dependencies: {
+      'nativescript-vue': '^2.2.2',
+      vuexp: '^0.3.0',
+      'vuexp-router': 'git://github.com/vuexp/vuexp-router.git#development',
```
Comment (brendtumi) [inline generator/index.js]: Using git path is not secure. Please give a compatible version number

File: generator/templates/advanced/src/components/HelloWorld.android.vue [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+</script>
+<%_ } _%>
+<!-- Add "scoped" attribute to limit CSS to this component only -->
+<%_ if (rootOptions.cssPreprocessor) { _%>
+<%_   if (rootOptions.cssPreprocessor == 'sass' || rootOptions.cssPreprocessor == 'scss'  || rootOptions.cssPreprocessor == 'dart-sass' ) { _%>
+<%#   -------------------- IS Using sass, scss OR dart-sass -------------------- -%>
+<style scoped lang="scss">
+.message {
+  color: #42b983;
+  margin-top: 80px;
```
Comment (brendtumi) [inline generator/templates/advanced/src/components/HelloWorld.android.vue]: use unit() function, apply to all page.

File: generator/templates/advanced/src/components/HelloWorld.ios.vue [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+</script>
+<%_ } _%>
+<!-- Add "scoped" attribute to limit CSS to this component only -->
+<%_ if (rootOptions.cssPreprocessor) { _%>
+<%_   if (rootOptions.cssPreprocessor == 'sass' || rootOptions.cssPreprocessor == 'scss'  || rootOptions.cssPreprocessor == 'dart-sass' ) { _%>
+<%#   -------------------- IS Using sass, scss OR dart-sass -------------------- -%>
+<style scoped lang="scss">
+.message {
+  color: #42b983;
+  margin-top: 80px;
```
Comment (brendtumi) [inline generator/templates/advanced/src/components/HelloWorld.ios.vue]: use unit() function, apply to all page.

File: generator/templates/advanced/src/components/HelloWorld.native.vue [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+</script>
+<%_ } _%>
+<!-- Add "scoped" attribute to limit CSS to this component only -->
+<%_ if (rootOptions.cssPreprocessor) { _%>
+<%_   if (rootOptions.cssPreprocessor == 'sass' || rootOptions.cssPreprocessor == 'scss'  || rootOptions.cssPreprocessor == 'dart-sass' ) { _%>
+<%#   -------------------- IS Using sass, scss OR dart-sass -------------------- -%>
+<style scoped lang="scss">
+.message {
+  color: #42b983;
+  margin-top: 80px;
```
Comment (brendtumi) [inline generator/templates/advanced/src/components/HelloWorld.native.vue]: use unit() function, apply to all page.

File: generator/templates/advanced/src/components/HelloWorld.vue [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+<!-- Add "scoped" attribute to limit CSS to this component only -->
+<%_ if (rootOptions.cssPreprocessor) { _%>
+<%_   if (rootOptions.cssPreprocessor == 'sass' || rootOptions.cssPreprocessor == 'scss'  || rootOptions.cssPreprocessor == 'dart-sass' ) { _%>
+<%#   -------------------- IS Using sass, scss OR dart-sass -------------------- -%>
+<style scoped lang="scss">
+.message {
+  color: #42b983;
+  display: block;
+  margin: auto;
+  margin-top: 40px;
```
Comment (brendtumi) [inline generator/templates/advanced/src/components/HelloWorld.vue]: use unit() function, apply to all page.

File: generator/templates/advanced/src/main.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
+---
+extend: '@vue/cli-service/generator/template/src/main.js'
+replace:
+  - !!js/regexp /import Vue from 'vue'/
+  - !!js/regexp /import App from './App.vue'/
```
Comment (brendtumi) [inline generator/templates/advanced/src/main.js:5]: Remove .vue extension by default

File: generator/templates/advanced/src/main.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  - !!js/regexp /Vue.config.productionTip = false/
+  - !!js/regexp /h => h\(App\),/
+  - !!js/regexp /}\)\.\$mount\('#app'\)/
+---
+
+<%# REPLACE %>
+import Vue from 'vue';
+import registerComponents from './componentRegisterer';
+import VxpPlugin from "vuexp/src/core/plugins/index.js";
```
Comment (brendtumi) [inline generator/templates/advanced/src/main.js]: Remove .js extension by default

File: generator/templates/advanced/src/main.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+Vue.use(VxpPlugin);
+Vue.mixin(AttributeMixin);
+Vue.mixin(GestureMixin);
+
+import "./styles/advanced.scss";
+<%# END_REPLACE %>
+
+<%# REPLACE %>
+import App from '~/App.vue';
```
Comment (brendtumi) [inline generator/templates/advanced/src/main.js]: Remove .vue extension by default

File: generator/templates/advanced/src/main.native.js [hunk 10]
```diff
@@ -0,0 +1,22 @@
+import Vue from 'nativescript-vue';
+import registerComponents from './componentRegisterer';
+import router from './router';
+
+import App from './App.vue';
```
Comment (brendtumi) [inline generator/templates/advanced/src/main.native.js]: Remove .vue extension by default

File: generator/templates/advanced/src/router.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
+---
+extend: '@vue/cli-service/generator/router/template/src/router.js'
+replace:
+  - !!js/regexp /import Vue from 'vue'/
+  - !!js/regexp /import Router from 'vue-router'/
+  - !!js/regexp /Vue.use\(Router\)/
+  - !!js/regexp /export default new Router\(\{/
+  - !!js/regexp /import Home from '\./views/Home.vue'/
```
Comment (brendtumi) [inline generator/templates/advanced/src/router.js:8]: Remove .vue extension by default

File: generator/templates/advanced/src/styles/style-one.scss [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+  height: 3em;
+  width: 100%;
+  top: 0px;
+  left: 0px;
+  margin: auto;
+  list-style: none;
+
+  display: flex;
+  align-items: center;
+  padding: 0 10px;
```
Comment (brendtumi) [inline generator/templates/advanced/src/styles/style-one.scss]: use unit() function, apply to all page.

File: generator/templates/advanced/src/styles/style-one.less [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+  height: 3em;
+  width: 100%;
+  top: 0px;
+  left: 0px;
+  margin: auto;
+  list-style: none;
+
+  display: flex;
+  align-items: center;
+  padding: 0 10px;
```
Comment (brendtumi) [inline generator/templates/advanced/src/styles/style-one.less]: use unit() function, apply to all page.

File: generator/templates/advanced/src/views/Home.vue [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+      this.$refs.sideLayout.showDrawerPanel();
+    },
+  }
+};
+</script>
+<%_ } else { _%>
+<%# -------------------- Is Using TypeScript  -------------------- -%>
+<script lang="ts">
+import { Component, Vue } from 'vue-property-decorator';
+import HelloWorld from 'components/HelloWorld.vue';
```
Comment (brendtumi) [inline generator/templates/advanced/src/views/Home.vue]: Remove .vue extension by default

File: generator/templates/simple/src/components/HelloWorld.android.vue [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+<%_ if (rootOptions.cssPreprocessor) { _%>
+<%_   if (rootOptions.cssPreprocessor == 'sass' || rootOptions.cssPreprocessor == 'scss'  || rootOptions.cssPreprocessor == 'dart-sass' ) { _%>
+<%#   -------------------- IS Using sass, scss OR dart-sass -------------------- -%>
+<style scoped lang="scss">
+.message {
+  color: #42b983;
+}
+
+image {
+  margin-top: 20px;
```
Comment (brendtumi) [inline generator/templates/simple/src/components/HelloWorld.android.vue]: use unit() function, apply to all page.

File: generator/templates/simple/src/components/HelloWorld.ios.vue [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+<%_ if (rootOptions.cssPreprocessor) { _%>
+<%_   if (rootOptions.cssPreprocessor == 'sass' || rootOptions.cssPreprocessor == 'scss'  || rootOptions.cssPreprocessor == 'dart-sass' ) { _%>
+<%#   -------------------- IS Using sass, scss OR dart-sass -------------------- -%>
+<style scoped lang="scss">
+.message {
+  color: #42b983;
+}
+
+image {
+  margin-top: 20px;
```
Comment (brendtumi) [inline generator/templates/simple/src/components/HelloWorld.ios.vue]: use unit() function, apply to all page.

File: generator/templates/simple/src/components/HelloWorld.native.vue [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+<%_ if (rootOptions.cssPreprocessor) { _%>
+<%_   if (rootOptions.cssPreprocessor == 'sass' || rootOptions.cssPreprocessor == 'scss'  || rootOptions.cssPreprocessor == 'dart-sass' ) { _%>
+<%#   -------------------- IS Using sass, scss OR dart-sass -------------------- -%>
+<style scoped lang="scss">
+.message {
+  color: #42b983;
+}
+
+image {
+  margin-top: 20px;
```
Comment (brendtumi) [inline generator/templates/simple/src/components/HelloWorld.native.vue]: use unit() function, apply to all page.

File: generator/templates/simple/src/main.js [hunk 18]
```diff
@@ -0,0 +1,31 @@
+---
+extend: '@vue/cli-service/generator/template/src/main.js'
+replace:
+  - !!js/regexp /import Vue from 'vue'/
+  - !!js/regexp /import App from './App.vue'/
```
Comment (brendtumi) [inline generator/templates/simple/src/main.js:5]: Remove .vue extension by default

File: generator/templates/simple/src/main.js [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+  - !!js/regexp /}\)\.\$mount\('#app'\)/
+---
+
+<%# REPLACE %>
+import Vue from 'vue';
+import registerComponents from './componentRegisterer';
+<%# END_REPLACE %>
+
+<%# REPLACE %>
+import App from '~/App.vue';
```
Comment (brendtumi) [inline generator/templates/simple/src/main.js]: Remove .vue extension by default

File: generator/templates/simple/src/main.native.js [hunk 20]
```diff
@@ -0,0 +1,20 @@
+import Vue from 'nativescript-vue';
+import registerComponents from './componentRegisterer';
+import router from './router';
+
+import App from './App.vue';
```
Comment (brendtumi) [inline generator/templates/simple/src/main.native.js]: Remove .vue extension by default

File: generator/templates/simple/src/router.js [hunk 21]
```diff
@@ -0,0 +1,53 @@
+---
+extend: '@vue/cli-service/generator/router/template/src/router.js'
+replace:
+  - !!js/regexp /import Vue from 'vue'/
+  - !!js/regexp /import Router from 'vue-router'/
+  - !!js/regexp /Vue.use\(Router\)/
+  - !!js/regexp /export default new Router\(\{/
+  - !!js/regexp /import Home from '\./views/Home.vue'/
```
Comment (brendtumi) [inline generator/templates/simple/src/router.js:8]: Remove .vue extension by default

File: generator/templates/simple/src/router.js [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+<%# REPLACE %>
+Vue.use(Router);
+<%# END_REPLACE %>
+
+<%# REPLACE %>
+export const options = {
+<%# END_REPLACE %>
+
+<%# REPLACE %>
+import Home from '~/views/Home.vue';
```
Comment (brendtumi) [inline generator/templates/simple/src/router.js]: Remove .vue extension by default

File: generator/templates/simple/src/styles/style-one.less [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+  height: 3em;
+  width: 100%;
+  top: 0px;
+  left: 0px;
+  margin: auto;
+  list-style: none;
+
+  display: flex;
+  align-items: center;
+  padding: 0 10px;
```
Comment (brendtumi) [inline generator/templates/simple/src/styles/style-one.less]: use unit() function, apply to all page.

File: generator/templates/simple/src/styles/style-one.scss [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+  height: 3em;
+  width: 100%;
+  top: 0px;
+  left: 0px;
+  margin: auto;
+  list-style: none;
+
+  display: flex;
+  align-items: center;
+  padding: 0 10px;
```
Comment (brendtumi) [inline generator/templates/simple/src/styles/style-one.scss]: use unit() function, apply to all page.

File: generator/templates/simple/src/views/Home.vue [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+      this.$router.push('/about');
+    }
+  }
+};
+</script>
+<%_ } else { _%>
+<%# -------------------- Is Using TypeScript  -------------------- -%>
+<script lang="ts">
+import { Component, Vue } from 'vue-property-decorator';
+import HelloWorld from 'components/HelloWorld.vue';
```
Comment (brendtumi) [inline generator/templates/simple/src/views/Home.vue]: Remove .vue extension by default

File: generator/templates/vue-sfc-template.vue [hunk 26]
```diff
@@ -0,0 +1,54 @@
+<%_ if (rootOptions.router) { _%>
+<%# -------------------- IS Using vue-router  -------------------- -%>
+<template web></template>
+<template native>
```
Comment (brendtumi) [inline generator/templates/vue-sfc-template.vue]: We don't have this support?

File: ui.js [hunk 27]
```diff
@@ -0,0 +1,19 @@
+/* eslint-disable no-unused-vars */
+module.exports = (api) => {
+  api.describeConfig({
+    // Unique ID for the config
+    id: 'org.nativescript-vue.plugin.cli',
```
Comment (brendtumi) [inline ui.js]: This is need to be updated.

**Review by brendtumi — CHANGES_REQUESTED** [review]

File: generator/index.js (hunk 1 above)
Comment (mylmz10) [reply generator/index.js]: Removed unnecessary packages

File: generator/index.js (hunk 2 above)
Comment (mylmz10) [reply generator/index.js]: It will be changed on release

File: generator/templates/advanced/src/main.js (hunk 8 above)
Comment (mylmz10) [reply generator/templates/advanced/src/main.js]: Removed

File: generator/templates/advanced/src/main.js (hunk 7 above)
Comment (mylmz10) [reply generator/templates/advanced/src/main.js:5]: It is the rule of replacing.

File: generator/templates/advanced/src/main.js (hunk 9 above)
Comment (mylmz10) [reply generator/templates/advanced/src/main.js]: Removed

File: generator/templates/advanced/src/main.native.js (hunk 10 above)
Comment (mylmz10) [reply generator/templates/advanced/src/main.native.js]: Removed

File: generator/templates/advanced/src/router.js (hunk 11 above)
Comment (mylmz10) [reply generator/templates/advanced/src/router.js:8]: It is the rule of replacing.

File: generator/templates/advanced/src/views/Home.vue (hunk 14 above)
Comment (mylmz10) [reply generator/templates/advanced/src/views/Home.vue]: Removed

File: generator/templates/simple/src/main.js (hunk 18 above)
Comment (mylmz10) [reply generator/templates/simple/src/main.js:5]: It is the rule of replacing

File: generator/templates/simple/src/main.js (hunk 19 above)
Comment (mylmz10) [reply generator/templates/simple/src/main.js]: Removed

File: generator/templates/simple/src/main.native.js (hunk 20 above)
Comment (mylmz10) [reply generator/templates/simple/src/main.native.js]: Removed

File: generator/templates/simple/src/router.js (hunk 21 above)
Comment (mylmz10) [reply generator/templates/simple/src/router.js:8]: It is the rule of replacing

File: generator/templates/simple/src/router.js (hunk 22 above)
Comment (mylmz10) [reply generator/templates/simple/src/router.js]: Removed

File: generator/templates/simple/src/views/Home.vue (hunk 25 above)
Comment (mylmz10) [reply generator/templates/simple/src/views/Home.vue]: Removed

File: generator/templates/vue-sfc-template.vue (hunk 26 above)
Comment (osmanertem) [reply generator/templates/vue-sfc-template.vue]: done

File: ui.js (hunk 27 above)
Comment (osmanertem) [reply ui.js]: done

File: generator/templates/advanced/src/components/HelloWorld.android.vue (hunk 3 above)
Comment (osmanertem) [reply generator/templates/advanced/src/components/HelloWorld.android.vue]: done

File: generator/templates/advanced/src/components/HelloWorld.ios.vue (hunk 4 above)
Comment (osmanertem) [reply generator/templates/advanced/src/components/HelloWorld.ios.vue]: done

File: generator/templates/advanced/src/components/HelloWorld.native.vue (hunk 5 above)
Comment (osmanertem) [reply generator/templates/advanced/src/components/HelloWorld.native.vue]: done

File: generator/templates/advanced/src/components/HelloWorld.vue (hunk 6 above)
Comment (osmanertem) [reply generator/templates/advanced/src/components/HelloWorld.vue]: done

File: generator/templates/advanced/src/styles/style-one.scss (hunk 12 above)
Comment (osmanertem) [reply generator/templates/advanced/src/styles/style-one.scss]: done

File: generator/templates/advanced/src/styles/style-one.less (hunk 13 above)
Comment (osmanertem) [reply generator/templates/advanced/src/styles/style-one.less]: done

File: generator/templates/simple/src/components/HelloWorld.android.vue (hunk 15 above)
Comment (osmanertem) [reply generator/templates/simple/src/components/HelloWorld.android.vue]: done

File: generator/templates/simple/src/components/HelloWorld.ios.vue (hunk 16 above)
Comment (osmanertem) [reply generator/templates/simple/src/components/HelloWorld.ios.vue]: done

File: generator/templates/simple/src/components/HelloWorld.native.vue (hunk 17 above)
Comment (osmanertem) [reply generator/templates/simple/src/components/HelloWorld.native.vue]: done

File: generator/templates/simple/src/styles/style-one.less (hunk 23 above)
Comment (osmanertem) [reply generator/templates/simple/src/styles/style-one.less]: done

File: generator/templates/simple/src/styles/style-one.scss (hunk 24 above)
Comment (osmanertem) [reply generator/templates/simple/src/styles/style-one.scss]: done

**Pull request merged**